File: app/src/main/java/soup/movie/ui/home/HomeListAdapter.java
package soup.movie.ui.home;

import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindColor;
import butterknife.BindView;
import butterknife.ButterKnife;
import soup.movie.R;
import soup.movie.data.soup.model.Movie;
import soup.movie.data.utils.MovieUtil;
import soup.movie.ui.detail.DetailActivity;
import soup.movie.ui.util.ImageUtil;
import soup.movie.util.ListUtil;

class HomeListAdapter extends RecyclerView.Adapter<HomeListAdapter.ViewHolder> {

    private final Activity mHost;

    private List<Movie> mItems = new ArrayList<>();

    @BindColor(R.color.green)
    int mGreenColor;

    @BindColor(R.color.blue)
    int mBlueColor;

    @BindColor(R.color.amber)
    int mAmberColor;

    @BindColor(R.color.red)
    int mRedColor;

    HomeListAdapter(Activity host) {
        mHost = host;
        ButterKnife.bind(this, host);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_movie_for_home, parent, false);
        ViewHolder holder = new ViewHolder(view);
        holder.itemView.setOnClickListener(v -> {
            Intent intent = new Intent(mHost, DetailActivity.class);
            MovieUtil.saveTo(intent, mItems.get(holder.getAdapterPosition()));
            ActivityOptions options =
                    ActivityOptions.makeSceneTransitionAnimation(mHost,
                            Pair.create(holder.mBackground, mHost.getString(R.string.transition_background)),
                            Pair.create(holder.mPosterView, mHost.getString(R.string.transition_poster)),
                            Pair.create(holder.mTitleView, mHost.getString(R.string.transition_title)),
                            Pair.create(holder.mAgeView, mHost.getString(R.string.transition_age)),
                            Pair.create(holder.mEggView, mHost.getString(R.string.transition_egg)),
//                            Pair.create(holder.mFavoriteButton, mHost.getString(R.string.transition_favorite)),
                            Pair.create(holder.mShareButton, mHost.getString(R.string.transition_share)));
            mHost.startActivity(intent, options.toBundle());
        });
        return holder;
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        Movie item = mItems.get(position);
        ImageUtil.loadAsync(mHost, holder.mPosterView, item.getPosterUrl());
        holder.mTitleView.setText(item.getTitle());
        holder.mEggView.setText(item.getEgg());
        setAgeText(holder.mAgeView, item.getAge());
    }

    private void setAgeText(TextView ageView, String age) {
        switch (age) {
            case "전체 관람가":
                ageView.setText("전체");
                ageView.setBackgroundTintList(ColorStateList.valueOf(mGreenColor));
                break;
            case "12세 관람가":
                ageView.setText("12");
                ageView.setBackgroundTintList(ColorStateList.valueOf(mBlueColor));
                break;
            case "15세 관람가":
                ageView.setText("15");
                ageView.setBackgroundTintList(ColorStateList.valueOf(mAmberColor));
                break;
            case "청소년관람불가":
                ageView.setText("청불");
                ageView.setBackgroundTintList(ColorStateList.valueOf(mRedColor));
                break;
            default:
                ageView.setVisibility(View.GONE);
        }
    }

    @Override
    public int getItemCount() {
        return ListUtil.size(mItems);
    }

    void updateList(List<Movie> newItems) {
        DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                return ListUtil.size(mItems);
            }

            @Override
            public int getNewListSize() {
                return ListUtil.size(newItems);
            }

            @Override
            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                return false;
            }

            @Override
            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                return false;
            }
        }, false);
        mItems = newItems;
        result.dispatchUpdatesTo(this);
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.background)
        View mBackground;
        @BindView(R.id.movie_poster)
        ImageView mPosterView;
        @BindView(R.id.primary_text)
        TextView mTitleView;
        @BindView(R.id.age_icon)
        TextView mAgeView;
        @BindView(R.id.sub_text2)
        TextView mEggView;
        @BindView(R.id.favorite_button)
        View mFavoriteButton;
        @BindView(R.id.share_button)
        View mShareButton;

        ViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
            mFavoriteButton.setOnClickListener(v -> {});
            mShareButton.setOnClickListener(v -> {});
        }
    }
}

File: app/src/main/java/soup/movie/ui/main/home/HomeListAdapter.kt
package soup.movie.ui.main.home

import android.util.Pair
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.ext.AlwaysDiffCallback
import kotlinx.android.synthetic.main.item_movie.view.*
import soup.movie.R
import soup.movie.data.helper.isBest
import soup.movie.data.helper.isDDay
import soup.movie.data.helper.isNew
import soup.movie.data.model.Movie
import soup.movie.ui.databinding.DataBindingListAdapter
import soup.movie.ui.databinding.DataBindingViewHolder
import soup.movie.util.consume
import soup.movie.util.setOnDebounceClickListener
import soup.movie.util.showToast
import soup.movie.util.with

class HomeListAdapter(
    private val listener: (Movie, Array<Pair<View, String>>) -> Unit
) : DataBindingListAdapter<Movie>(AlwaysDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DataBindingViewHolder<Movie> {
        return super.onCreateViewHolder(parent, viewType).apply {
            itemView.setOnDebounceClickListener {
                val index = adapterPosition
                if (adapterPosition in 0..itemCount) {
                    val movie: Movie = getItem(index)
                    listener(movie, createSharedElements(movie))
                }
            }
            itemView.setOnLongClickListener {
                consume {
                    val index = adapterPosition
                    if (index in 0..itemCount) {
                        val movie: Movie = getItem(index)
                        it?.context?.showToast(movie.title)
                    }
                }
            }
        }
    }

    private fun DataBindingViewHolder<Movie>.createSharedElements(movie: Movie): Array<Pair<View, String>> {
        itemView.run {
            val sharedElements = mutableListOf(
                backgroundView with R.string.transition_background,
                posterView with R.string.transition_poster,
                ageBgView with R.string.transition_age_bg)
            if (movie.isNew()) {
                sharedElements.add(newView with R.string.transition_new)
            }
            if (movie.isBest()) {
                sharedElements.add(bestView with R.string.transition_best)
            }
            if (movie.isDDay()) {
                sharedElements.add(dDayView with R.string.transition_d_day)
            }
            return sharedElements.toTypedArray()
        }
    }

    override fun getItemViewType(position: Int) = R.layout.item_movie
}
